### PR TITLE
96 indroduce here doc

### LINKED ISSUE #96: Indroduce HERE_DOC
## Memo

``` bash
unset $HOME && cat << EOF
> $HOME
> EOF
# no characters displayed
```

So it is not appropriate to apply expand_word on parsing.
Parser do only create and write to a file for here document.
word_expand should be applied on executing.

### Specification

`cat << EOF`, `cat << "EOF"`, `cat << 'EOF'`
quote removal needed

``` bash
> $PWD
> EOF
/home/user/clang/minishell
```
parameter expansion needed

### PR DESCRIPTION
**Pull request by rask24**
fix #96

- **Implement basic here document**
- **Update heredoc tmpfile**
- **Update handle_heredoc**
- **Fix test_ast**
- **Add basic e2e tests for heredoc**

## Summary by CodeRabbit

- **New Features**
	- Introduced functionality for managing heredoc input, allowing users to input multiple lines until a specified delimiter.
	- Added support for handling heredoc file descriptors within the redirect management logic.

- **Bug Fixes**
	- Enhanced safety checks for closing heredoc file descriptors to prevent invalid operations.

- **Tests**
	- Implemented end-to-end tests to validate heredoc functionality in a shell session, ensuring expected behavior with both single and multiple inputs.
	- Updated unit tests to reflect changes in the construction of redirect information, ensuring consistency across test cases.

### CODE REVIEW

File: src/parser/handle_heredoc.c [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+		if (ft_strcmp(line, delimiter) == 0)
+		{
+			free(line);
+			break ;
+		}
+		heredoc_size += ft_strlen(line) + 1;
+		ft_lstadd_back(&input_list, ft_xlstnew(line));
+	}
+	return (open_heredoc(input_list, heredoc_size));
+}
```
[the remark this answers is not in the document]
Comment (rask24) [reply src/parser/handle_heredoc.c:96]: I'll handle this in another issue.

File: src/parser/handle_heredoc.c [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+	unlink(tmpfile);
+	return (fd);
+}
+
+static int	open_heredoc(t_list *input_list, size_t heredoc_size)
+{
+	int	pipe_fd[2];
+
+	if (heredoc_size == 0)
+		return (open("/dev/null", O_RDONLY));
```
[the remark this answers is not in the document]
Comment (rask24) [reply src/parser/handle_heredoc.c:78]: Exactly.

File: src/parser/handle_heredoc.c [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+			print_heredoc_warning(delimiter);
+			break ;
+		}
+		if (ft_strcmp(line, delimiter) == 0)
+		{
+			free(line);
+			break ;
+		}
+		heredoc_size += ft_strlen(line) + 1;
+		ft_lstadd_back(&input_list, ft_xlstnew(ft_xstrdup(line)));
```
Comment (alex216) [inline src/parser/handle_heredoc.c:116]: do we need to free `list` here?
```suggestion
		ft_lstadd_back(&input_list, ft_xlstnew(line));
```

or

```suggestion
		ft_lstadd_back(&input_list, ft_xlstnew(ft_xstrdup(line)));
		free(line);
```

**Review by alex216 — APPROVED** [review]
@rask24 ok!

**Pull request merged**